pkg/odo/cli/component/common_push.go

```diff
@@ -330,3 +332,75 @@ func filterIgnores(filesChanged, filesDeleted, absIgnoreRules []string) (filesCh
 	}
 	return filesChangedFiltered, filesDeletedFiltered
 }
+
+// retrieveKubernetesDefaultNamespace tries to retrieve the current active namespace
+// to set as a default namespace
+func retrieveKubernetesDefaultNamespace() (string, error) {
+	// Get current active namespace
+	client, err := kclient.New()
+	if err != nil {
+		return "", err
+	}
+	return client.Namespace, nil
+}
+
+// retrieveCmdNamespaceDefaultIfNotAvailable retrieves the the namespace from project or namespace, if neither of those are available
+// we revert to the default namespace available from Kubernetes
+func retrieveCmdNamespaceDefaultIfNotAvailable(cmd *cobra.Command) (string, error) {
+	var componentNamespace string
+	var err error
+
+	// For "odo create" check to see if --project has been passed.
+	if cmd.Flags().Changed("project") {
+		componentNamespace, err = cmd.Flags().GetString("project")
+		if err != nil {
+			return "", err
+		}
+	} else if cmd.Flags().Changed("namespace") {
+		// For "odo push" check to see if project has been passed
+		componentNamespace, err = cmd.Flags().GetString("namespace")
+		if err != nil {
+			return "", err
+		}
+	} else {
+		componentNamespace, err = retrieveKubernetesDefaultNamespace()
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return componentNamespace, nil
+}
+
+// gatherName parses the Devfile retrieves an appropriate name in two ways.
+// 1. If metadata.name exists, we use it
+// 2. If metadata.name does NOT exist, we use the folder name where the devfile.yaml is located
+func gatherName(devfilePath string) (string, error) {
+
+	// Parse the Devfile
+	devObj, err := devfile.Parse(devfilePath)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to parse devfile while gathering the component name")
+	}
```

Comment:
Can we parse this devfile before and pass into the function? Because we seem to call it later at https://github.com/openshift/odo/blob/e28812a34d4a2443d1e816141c401e3562a5c6ae/pkg/odo/cli/component/devfile.go#L58. Maybe parse and pass into these functions?

Reply:
@mik-dass Within pkg/odo/cli/component/push.go we don't parse before we gather the name, so we have to keep the parser in here: https://github.com/cdrage/odo/blob/devfile-push-without-odo-dir/pkg/odo/cli/component/push.go#L115

Reply:
But this file later calls https://github.com/openshift/odo/blob/e28812a34d4a2443d1e816141c401e3562a5c6ae/pkg/odo/cli/component/devfile.go#L37
which later calls https://github.com/openshift/odo/blob/e28812a34d4a2443d1e816141c401e3562a5c6ae/pkg/odo/cli/component/devfile.go#L58

So maybe we can refactor and call Parse() once and use the data in these functions probably by storing in the PushOptions?

Reply:
can we do that in a follow up? This PR is long overdue and i can record this cleanup in a separate issue.

Reply:
https://github.com/openshift/odo/issues/3520